### PR TITLE
Add @DisabledUntil annotation and extension (#366)

### PR DESCRIPTION
**Pull request by gdiegel**
Proposed commit message:

```
Create new @DisabledUntil extension. (#366 / #529)

Add the @DisabledUntil annotation and corresponding extension. The
implementation uses a required 'untilDate' parameter and may also be
inherited via the @Inherited meta-annotation. Test annotated with
@DisabledUntil will be skipped if the date is in the future, i.e.
today plus n days where n > 0. If the untilDate parameter happens to be
today or any day in the past, the test will be executed but a warning
report entry will be published. Add tests for the extension. Add site
documentation. Add contributer entry in README.

Resolves: #366
PR: #529
```
---
**PR checklist**

The following checklist shall help the PR's author, the reviewers and maintainers to ensure the quality of this project.
It is based on our contributors guidelines, especially the ["writing code" section](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#writing-code).
It shall help to check for completion of the listed points.
If a point does not apply to the given PR's changes, the corresponding entry can be simply marked as done. 

Documentation (general)
* [ ] There is documentation (Javadoc and site documentation; added or updated)
* [ ] There is implementation information to describe _why_ a non-obvious source code / solution got implemented
* [ ] Site documentation has its own `.adoc` file in the `docs` folder, e.g. `docs/report-entries.adoc`
* [ ] Only one sentence per line (especially in `.adoc` files)
* [ ] Javadoc uses formal style, while sites documentation may use informal style

Documentation (new extension)
* [ ] The `docs/docs-nav.yml` navigation has an entry for the new extension
* [ ] The `package-info.java` contains information about the new extension

Code
* [ ] Code adheres to code style, naming conventions etc.
* [ ] Successful tests cover all changes
* [ ] There are checks which validate correct / false usage / configuration of a functionality and there are tests to verify those checks
* [ ] Tests use [AssertJ](https://assertj.github.io/doc/) or our own [PioneerAssert](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#assertions) (which are based on AssertJ)

Contributing
* [ ] A prepared commit message exists
* [ ] The list of contributions inside `README.md` mentions the new contribution (real name optional) 

---

I hereby agree to the terms of the [JUnit Pioneer Contributor License Agreement](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#junit-pioneer-contributor-license-agreement).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -115,6 +115,7 @@ The least we can do is to thank them and list some of their accomplishments here
 
 #### 2021
 
+* [Gabriel Diegel](https://github.com/gdiegel) contributed the `@DisabledUntil` extension in [Temporarily disable a test](https://junit-pioneer.org/docs/disabled-until/) (#366)
```
Comment (Michael1993) [inline README.md:118]: Please, follow lexicographic order.

File: docs/disabled-until.adoc [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    @Test
+    void test() {
+        // Test will be skipped if it's 2021-12-31 or earlier
+    }
+}
+----
+
+The `@DisabledUntil` annotation can only be used once per class or method.
+
+== Implementation
```
Comment (Michael1993) [inline docs/disabled-until.adoc:54]: What's the purpose of this section?

File: src/test/java/org/junitpioneer/internal/PioneerDateUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import java.time.LocalDate;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@DisplayName("Pioneer date utilities")
+class PioneerDateUtilsTest {
```
Comment (Michael1993) [inline src/test/java/org/junitpioneer/internal/PioneerDateUtilsTest.java:21]: These tests are not discovered by our build. Please rename this class as `PioneerDateUtilsTests`.

File: src/test/java/org/junitpioneer/jupiter/DisabledUntilExtensionTest.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junitpioneer.testkit.ExecutionResults;
+import org.junitpioneer.testkit.PioneerTestKit;
+import org.junitpioneer.testkit.assertion.PioneerAssert;
+
+@DisplayName("Tests for the DisabledUntil extension")
+class DisabledUntilExtensionTest {
```
Comment (Michael1993) [inline src/test/java/org/junitpioneer/jupiter/DisabledUntilExtensionTest.java:21]: These tests are not discovered by our build. Please rename this class as `DisabledUntilExtensionTests`.

**Review by Michael1993 — CHANGES_REQUESTED** [review]
Looks splendid, just some minor things that need to be polished.

File: docs/disabled-until.adoc [hunk 5]
```diff
@@ -0,0 +1,68 @@
+:page-title: Temporarily disable a test
+:page-description: Extends JUnit Jupiter with `@DisabledUntil` to temporarily disable a test.
+
+== Introduction
+
+It's sometimes useful to disable a test.
+One can imagine many reasons for doing so, maybe a remote dependency of the test is broken or not yet deployed.
+Maybe the test is still in development and unstable.
+
+The `@Disable` annotation is perfectly adequate to use in such cases, the test gets skipped and the test suite can terminate successfully.
```
[the remark this answers is not in the document]
Comment (gdiegel) [reply docs/disabled-until.adoc:10]: Done.

File: README.md (hunk 1 above)
Comment (gdiegel) [reply README.md:118]: Done.

File: src/test/java/org/junitpioneer/jupiter/DisabledUntilExtensionTest.java (hunk 4 above)
Comment (gdiegel) [reply src/test/java/org/junitpioneer/jupiter/DisabledUntilExtensionTest.java:21]: Done.

File: src/test/java/org/junitpioneer/internal/PioneerDateUtilsTest.java (hunk 3 above)
Comment (gdiegel) [reply src/test/java/org/junitpioneer/internal/PioneerDateUtilsTest.java:21]: Done.

**Review by Michael1993 — APPROVED** [review]
Looks great, this can be merged.

File: src/test/java/org/junitpioneer/jupiter/DisabledUntilExtensionTests.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+import org.junit.jupiter.api.Test;
+import org.junitpioneer.testkit.ExecutionResults;
+import org.junitpioneer.testkit.PioneerTestKit;
+import org.junitpioneer.testkit.assertion.PioneerAssert;
+
+@DisplayName("Tests for the DisabledUntil extension")
+class DisabledUntilExtensionTests {
+
+	@Test
+	@DisplayName("Should enable test without annotation")
```
[the remark this answers is not in the document]
Comment (gdiegel) [reply src/test/java/org/junitpioneer/jupiter/DisabledUntilExtensionTests.java:24]: I'm not sure that there's a clear answer to this question. If you consider these tests to be executable application specification, you'd need a test case for the use case "No annotation".

**Comment by nipafx** [comment]
Thank you very much for the contribution, @gdiegel, it turned out really great. I refactored a bunch of small things (like using `String::format` and statically importing assertions) and three more interesting ones:

* throw exception on unparsable date string, so if the user makes a mistake, we tell them instead of failing silently (although when a test that you assume is disabled because it fails, isn't, that's probably not so silent anyway 😉 )
* renamed `untilDate` to `date` because even though `untilDate` would be a better name for a field, as an annotation attribute where `@DisabledUntil` is always right in front of it, it reads a bit better (e.g. `@DisabledUntil(date = "...")` instead of `@DisabledUntil(untilDate = "...")`
* removed the utility class because I think it doesn't add much as each method was only called once

**Comment by nipafx** [comment]
Let me know what you think and if you're ok with the changes, we can merge this.

**Review by nipafx — APPROVED** [review]

**Review by Bukama — APPROVED** [review]

**Comment by gdiegel** [comment]
Hi @nipafx, thanks for your feedback. I’m ok with your changes, this may be merged.

**Pull request merged**